### PR TITLE
Safe Upload

### PR DESCRIPTION
**Pull request by arnohovhannisyan**
#### Short description of what this resolves:
This PR enables "safe" upload functionality.

#### Changes proposed in this pull request:

- Check if contents of gist are identical or newer to local settings and don't upload if they are (unless forceUpload is turned on)
- Implement `forceUpload`

**Fixes**: #350 #316 

#### How Has This Been Tested?

Tested by uploading to gist with contents older or identical to the gist.

#### Checklist:

- [x] I have read the [contribution](https://github.com/shanalikhan/code-settings-sync/blob/master/CONTRIBUTING.md#setup-extension-locally) guidelines.
- [x] My change requires a change to the documentation and GitHub Wiki.
- [ ] I have updated the documentation and Wiki accordingly.

### CODE REVIEW

File: src/service/github.service.ts [hunk 1]
```diff
@@ -131,6 +131,20 @@ export class GitHubService {
     return await this.github.gists.get({ gist_id: GIST });
   }
 
+  public async CheckIfGistIsNewer(
```
Comment (shanalikhan) [inline src/service/github.service.ts]: change to `IsGistNewer`

File: src/sync.ts [hunk 2]
```diff
@@ -82,6 +82,27 @@ export class Sync {
...
+          )
+        ) {
+          const message = await vscode.window.showInformationMessage(
+            localize("cmd.updateSettings.warning.gistNewer"),
+            "Yes"
+          );
+          if (message === "Yes") {
+            localConfig.extConfig.forceUpload = true;
+            await state.commons.SaveSettings(localConfig.extConfig);
+          } else {
```
Comment (shanalikhan) [inline src/sync.ts]: We need to show status bar message as
`Sync : Upload Process Cancelled` for sometime (maybe 3-5 seconds)

for the else part.

File: src/service/github.service.ts (hunk 1 above)
Comment (arnohovhannisyan) [reply src/service/github.service.ts]: Done

File: src/sync.ts (hunk 2 above)
Comment (arnohovhannisyan) [reply src/sync.ts]: Done

**Comment by shanalikhan** [comment]
> Tested by uploading to gist with contents older or identical to the gist.

Can you try with other scenario by pushing from one computer and then pushing from second computer ( without pulling ) , It should not upload as first computer has pushed the most updated settings.

**Comment by arnohovhannisyan** [comment]
@shanalikhan I have just tested that and it works.

**Comment by shanalikhan** [comment]
image


Gist LINK : https://gist.github.com/shanalikhan/e86911443d3f81be6a66699603aee6dc


It should not ask to force upload because lastUpload is Null and in cloudSettings settings date is old.

**Comment by arnohovhannisyan** [comment]
@shanalikhan This will be fixed in #937.

**Comment by shanalikhan** [comment]
Can you fix it in `v3.4.0` branch? We can ship it in this week.

**Comment by shanalikhan** [comment]
For #316 
I'm uploading the same settings again and again and it is uploading.
It should not upload the settings on second upload and prompt that settings are the same and ask user to enable forceUpload to perform upload.

**Comment by shanalikhan** [comment]
> Can you test autoUpload scenarios at your end by testing on package and install locally.

Any update on this ?

So the PR you have send will not allow user to upload the settings second time if there is not any difference between first upload content ?

**Comment by arnohovhannisyan** [comment]
If there is no difference between remote and local settings, a message box will be shown to the user informing them about it and giving them the option to enable `forceUpload`. If they enable it, the upload will continue.

**Pull request merged**